### PR TITLE
Dashboard Routing and Build Optimizations

### PR DESCRIPTION
**Pull request by RedLeader16**
# Description
`nginx` location routing is not gracefully handling `next.js` dynamic routing in a graceful and scalable way.
By using an internal node server and proxying dashboard requests to it, all `next.js` features are enabled without special configurations. 

`next.js` [output file tracing](https://nextjs.org/docs/advanced-features/output-file-tracing) also helps reduce the overall featureform image size (`~4.2GBs` vs. `~1.9GBs`).

More importantly, the production dashboard operates exactly as our local node development environments do. Currently unexpected errors can occur once our code hits production that we might miss in local dev. 

[Proposal Document](https://docs.google.com/document/d/1sv8zJPiaRCjBfmJ4u8rwWkzZ7z1EGAGWycfYsUHkZxI/edit#heading=h.gjdgxs)
[Excalidraw Design](https://link.excalidraw.com/readonly/ulrwuOwtBjhcvXeFFqtz)

This merge request will remain in **_draft_** for the next couple of weeks, to enable thorough testing + evaluation.

Testing Schedule:
- [x] Local Mode
- [X] Docker Mode
- [x] K8S / Minikube Mode
- [x] Bug Bash
 

## Type of change
Routing architecture + docker build change. 

### Does this correspond to an open issue?

https://github.com/featureform/featureform/issues/774

### Select type(s) of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: dashboard/Dockerfile [hunk 1]
```diff
@@ -1,18 +1,40 @@
-# RUN docker build . -t dashboard
```
Comment (aolfat) [inline dashboard/Dockerfile:1]: I'd throw this into ChatGPT just to see if it has any more optimizations:

I'd use this prompt:

"if you were an expert on best practices for a dockerfile, how would you make this one more efficient and abide by better conventions"

File: Dockerfile [hunk 2]
```diff
@@ -84,9 +110,16 @@ COPY nginx.conf /etc/nginx/nginx.conf
...
+RUN apt install nodejs
+
 ENV SERVING_PORT="8082"
 ENV SERVING_HOST="0.0.0.0"
 ENV ETCD_ARCH=""
 
 EXPOSE 7878
-CMD ["/usr/bin/supervisord"]
\ No newline at end of file
+EXPOSE 80
```
Comment (aolfat) [inline Dockerfile:123]: changing the exposed port doesn't break anything?

File: dashboard/Dockerfile (hunk 1 above)
Comment (aolfat) [reply dashboard/Dockerfile:1]: same with the above dockerfile

File: Dockerfile (hunk 2 above)
Comment (RedLeader16) [reply Dockerfile:123]: double checking

File: dashboard/Dockerfile (hunk 1 above)
Comment (RedLeader16) [reply dashboard/Dockerfile:1]: that file you sent had a couple of good optimizations. will add soon + test.

File: Dockerfile (hunk 2 above)
Comment (RedLeader16) [reply Dockerfile:123]: it doesn't break anything that I've tested. we expose 7878 + 80 when we run FF in the CLI, this makes it explicit.

File: dashboard/Dockerfile (hunk 1 above)
Comment (RedLeader16) [reply dashboard/Dockerfile:1]: tried a few of them but running into a couple of errors at build time (with npm, etc.)
I'll stick to the current implementation for now. using next.js [docs](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) as inspiration.

**Review by ahmadnazeri — APPROVED** [review]

**Pull request merged**